feat(ctb): `CONTRIBUTING.md` (#6790)
ethereum-optimism/optimism: packages/contracts-bedrock/CONTRIBUTING.md

# Contributing to CONTRIBUTING.md

First off, thanks for taking the time to contribute!

> We welcome and appreciate all kinds of contributions. Refer to the Table of Contents to discover various ways you can assist, as well as the procedures we follow for each type. Before you contribute, kindly review the appropriate section; this will streamline the process for both maintainers and contributors. We're excited to see your contributions! 🎉

Reviewer: Might be worth mentioning that large contributions should consider opening an issue first so we can make sure they don't do a lot of work for a contribution we ultimately would reject

> First off, thanks for taking the time to contribute!
> 
> We welcome and appreciate all kinds of contributions. Refer to the Table of Contents to discover various ways you can assist, as well as the procedures we follow for each type. Before you contribute, kindly review the appropriate section; this will streamline the process for both maintainers and contributors. We're excited to see your contributions! 🎉

Reviewer 2: Adding an explicit suggestion to @roninjin10's comment:

```suggestion
First off, thanks for taking the time to contribute!

We welcome and appreciate all kinds of contributions. We ask that before contributing you please review the procedures for each type of contribution available in the [Table of Contents](#table-of-contents). This will streamline the process for both maintainers and contributors. To find ways to contribute, view the [I Want To Contribute](#i-want-to-contribute) section below. Larger contributions should [open an issue](https://github.com/ethereum-optimism/optimism/issues/new) before implementation to ensure changes don't go to waste.

We're excited to work with you and your contributions to scaling Ethereum!
```

## Table of Contents

- [I Have a Question](#i-have-a-question)
- [I Want To Contribute](#i-want-to-contribute)
- [Reporting Bugs](#reporting-bugs)
- [Suggesting Enhancements](#suggesting-enhancements)
- [Your First Code Contribution](#your-first-code-contribution)
- [Improving The Documentation](#improving-the-documentation)
- [Deploying on Devnet](#deploying-on-devnet)
- [Tools](#tools)

## I Have a Question

> **Note**
> Before making an issue, please read the documentation and search the issues to see if your question has already been answered.

> If you have any questions about the smart contracts, please feel free to ask them in the Optimism discord developer channels or in the issues of the monorepo.

Reviewer 2: `monorepo` is a bit non-specific
```suggestion
If you have any questions about the smart contracts, please feel free to ask them in the Optimism discord developer channels or create a new detailed issue.
```

## I Want To Contribute

### Reporting Bugs

**Any and all bug reports on production smart contract code should be submitted privately to the Optimism team so that we can mitigate the issue before it is exploited. Please see our security policy document [here](https://github.com/ethereum-optimism/.github/blob/master/SECURITY.md).**

### Suggesting Enhancements

#### Before Submitting an Enhancement

- Read the documentation and the smart contracts themselves to see if the feature already exists.
- Perform a search in the issues to see if the enhancement has already been suggested. If it has, add a comment to the existing issue instead of opening a new one.

#### How Do I Submit a Good Enhancement Suggestion?

Enhancement suggestions are tracked as [GitHub issues](/issues).

- Use a **clear and descriptive title** for the issue to identify the suggestion.

> - Provide a **step-by-step description of the suggested enhancement** in as many details as possible.

Reviewer 2: nit
```suggestion
- Provide a **step-by-step** description of the suggested enhancement in as many details as possible.
```

> - **Describe the current behavior** and **explain which behavior you expected to see instead** and why. At this point you can also tell which alternatives do not work for you.

Reviewer 2: nit
```suggestion
- Describe the **current** behavior and why the **intended** behavior you expected to see differs. At this point you can also tell which alternatives do not work for you.
```

> - **Explain why this enhancement would be useful** in Optimism's smart contracts. You may also want to point out the other projects that solved it better and which could serve as inspiration.

Reviewer 2: nit
```suggestion
- Explain why this enhancement would be useful in Optimism's smart contracts. You may also want to point out the other projects that solved it better and which could serve as inspiration.
```

1. `pnpm test` Runs the full `forge` test suite.
1  `pnpm gas-snapshot` Generates the gas snapshot for the smart contracts.
1. `pnpm semver-lock` Generates the semver lockfile.
1. `pnpm storage-snapshot` Generates the storage lockfile.
1. `pnpm autogen:invariant-docs` Generates the invariant test documentation.
1. `pnpm clean` Removes all build artifacts for `forge` and `go` compilations.
1. `pnpm validate-spacers` Validates the positions of the storage slot spacers.
1. `pnpm validate-deploy-configs` Validates the deployment configurations in `deploy-config`
1. `pnpm slither` Runs the slither static analysis tool on the smart contracts.
1. `pnpm lint` Runs the linter on the smart contracts and scripts.
1. `pnpm pre-pr` Runs most checks, generators, and linters prior to a PR. For most PRs, this is sufficient to pass CI if everything is in order.
1. `pnpm pre-pr:full` Runs all checks, generators, and linters prior to a PR.

### Improving The Documentation

Documentation improvements are more than welcome! If you see a typo or feel that a code comment describes something poorly or incorrectly, please submit a PR with a fix.

### Deploying on Devnet

> To deploy the smart contracts on a local devnet, run `make devnet-up` in the monorepo root.

Reviewer 2: ```suggestion
To deploy the smart contracts on a local devnet, run `make devnet-up` in the monorepo root. This will serve an rpc endpoint at localhost on port `8545` (qualified as: `127.0.0.1:8545`). When using `forge`, this is the default rpc url, so you shouldn't need to pass it in with the `--rpc-url 127.0.0.1:8545` flag.
```

Author: hmm I think this isn't accurate wrt the rpc url flag. Maybe we point to the devnet documentation? Gives more info than a sentence or two could.

Author: Talked offline, added a link to `devnet.md`

### Tools

#### Layout Locking

We use a system called "layout locking" as a safety mechanism to prevent certain contract variables from being moved to different storage slots accidentally.
To lock a contract variable, add it to the `layout-lock.json` file which has the following format:

```json
{
  "MyContractName": {
    "myVariableName": {
      "slot": 1,
      "offset": 0,
      "length": 32
    }
  }
}
```

> With the above config, the `validate-spacers` hardhat task will check that we have a contract called `MyContractName`, that the contract has a variable named `myVariableName`, and that the variable is in the correct position as defined in the lock file.

Reviewer 2: ```suggestion
With the above config, the `validate-spacers` script will check that we have a contract called `MyContractName`, that the contract has a variable named `myVariableName`, and that the variable is in the correct position as defined in the lock file.
```
